### DISCUSSION TITLE
atomWithLazy types and docs seem out of alignment

### DISCUSSION THREAD

**Discussion by trappar** [pmndrs/jotai#2711] [Bug report] [answered]
### Bug Description

The docs state:

> We can use atomWithLazy to create a primitive atom whose initial value will be computed at [first use in the store](https://jotai.org/docs/utilities/lazy#using-multiple-stores). After initialization, it will behave like a regular primitive atom (can be written to).

My interpretation (maybe I'm wrong) of this is that it would only be useful if the initialization function is async. For example

```typescript
import { useAtom } from 'jotai';
import { atomWithLazy } from 'jotai/utils';

const selectedItemsAtom = atomWithLazy(async () => {
  const response = await fetch('/api/recommended-items');
  const data = (await response.json()) as { items: string[] };

  return data.items;
});

export default function Page() {
  const [selectedItems, setSelectedItems] = useAtom(selectedItemsAtom);

  return (
    <div>
      Selected items: {selectedItems.join(', ')}
      <br />
      <button onClick={() => setSelectedItems([...selectedItems, 'foo'])}>
        Add additional item
      </button>
    </div>
  );
}
```

This code functions just fine, but causes a typescript error when calling `setSelectedItems`:
```
Argument of type 'string[]' is not assignable to parameter of type 'SetStateAction<Promise<string[]>>'
```

The setter expects the incoming type to be a `Promise<string[]>` instead of a `string[]`, which brings me back do the docs:

> After initialization, it will behave like a regular primitive atom (can be written to)

Which doesn't seem quite right.

I can work around it by updating the usage of the setter function to pass in a promise that resolves to the desired value like this:
```typescript
setSelectedItems(Promise.resolve([...selectedItems, 'foo']))
```

But that seems like a very odd thing to do.

**So is the code / docs wrong? Or am I?**

### Reproduction Link

Minimal repro included in description above.

**Comment by trappar** [comment]
### Additional questions

**Is my assumption that `atomWithLazy` would be used with an async initializer wrong?**

**Is the following a more idiomatic way to do this with jotai?**

```typescript
import { atom, useAtom } from 'jotai';

export const userSelectedItemsAtom = atom<undefined | string[]>(undefined);

const recommendedItemsAtom = atom(async () => {
  const response = await fetch('/api/recommended-items');
  const data = (await response.json()) as { items: string[] };

  return data.items;
});

const selectedItemsAtom = atom(
  (get) => {
    const value = get(userSelectedItemsAtom);

    return value ?? get(recommendedItemsAtom);
  },
  (_get, set, value: string[]) => {
    set(userSelectedItemsAtom, value);
  },
);
```

Seems like a lot more boilerplate just to get the same result...

**Comment by dai-shi** [comment]
> My interpretation (maybe I'm wrong) of this is that it would only be useful if the initialization function is async.

No, it's still useful for sync atoms. ex:
```js
atomWithLazy(() => new Set());
```

**Comment by dai-shi** [comment, marked as answer]
> This code functions just fine, but causes a typescript error

You need to do like this:

```ts
const selectedItemsAtom = atomWithLazy<Promise<string[]> | string[]>(async () => {
  const response = await fetch('/api/recommended-items');
  const data = (await response.json()) as { items: string[] };

  return data.items;
});
```

Note that this is not a specific issue/gocha with `atomWithLazy`.

It's the same with normal `atom`.
```ts
atom(Promise.resolve(0))
atom<Promise<number> | number>(Promise.resolve(0))
```
